provider/openstack/provider.go

```diff
@@ -1338,6 +1345,40 @@ func (e *Environ) StartInstance(ctx context.ProviderCallContext, args environs.S
 	}, nil
 }
 
+func (e *Environ) configureRootDisk(ctx context.ProviderCallContext, args environs.StartInstanceParams,
+	spec *instances.InstanceSpec, runOpts *nova.RunServerOpts) error {
+	rootDiskSource := "local"
+	if args.Constraints.HasRootDiskSource() {
+		rootDiskSource = *args.Constraints.RootDiskSource
+	}
+	rootDiskMapping := nova.BlockDeviceMapping{
+		BootIndex:  0,
+		UUID:       spec.Image.Id,
+		SourceType: "image",
+		// NB constraints.RootDiskSource in the case of OpenStack represents
+		// the type of block device to use. Either "local" to represent a local
+		// block device or "volume" to represent a block device from the cinder
+		// block storage service.
+		DestinationType:     rootDiskSource,
+		DeleteOnTermination: true,
+	}
+	switch rootDiskSource {
+	case "local":
+		runOpts.ImageId = spec.Image.Id
+	case "volume":
+		size := spec.InstanceType.RootDisk
+		if args.Constraints.HasRootDisk() {
```

Comment:
don't we set root disk constraint to nil if root disk source == "volume" elsewhere?

Reply:
RootDisk is only set to nil in the findInstanceSpec to prevent matching instance flavors on RootDisk, due to RootDisk having a different meaning when RootDiskSource=volume. Otherwise when someone specifies RootDiskSource=volume and RootDisk=1TB, we don't want to match a instance type that has 1TB of local storage, just we want the root disk cinder volume to have 1TB.